refactor prepare_mask_and_masked_image with VaeImageProcessor

yiyixuxu opened the pull request:
first attempt to refactor inpainting pipelines with VaeImageProcessor
lots of tests need to be added


This example work as expected after refactoring 

```python
import PIL
import requests
import torch
from io import BytesIO

from diffusers import StableDiffusionInpaintPipeline


def download_image(url):
    response = requests.get(url)
    return PIL.Image.open(BytesIO(response.content)).convert("RGB")


img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = download_image(img_url).resize((512, 512))
mask_image = download_image(mask_url).resize((512, 512))

generator = torch.Generator(device="cuda").manual_seed(0)

pipe = StableDiffusionInpaintPipeline.from_pretrained(
    "runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16,
)
pipe = pipe.to("cuda")

prompt = "Face of a yellow cat, high resolution, sitting on a park bench"
image = pipe(prompt=prompt, image=init_image, mask_image=mask_image, generator=generator).images[0]
image.save("yellow_cat.png")
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4444). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:
```diff
@@ -858,9 +866,17 @@ def __call__(
+
+        mask = self.mask_processor.preprocess(mask_image, height=height, width=width)
+        # binarize mask
+        mask[mask < 0.5] = 0
+        mask[mask > 0.5] = 1
```
```suggestion
        mask[mask < 0.5] = 0
        mask[mask > 0.5] = 1
```

shouldn't we do this in the preprocess function?

patrickvonplaten commented on src/diffusers/image_processor.py:
```diff
@@ -52,8 +52,16 @@ def __init__(
         resample: str = "lanczos",
         do_normalize: bool = True,
         do_convert_rgb: bool = False,
+        do_convert_grayscale: bool = False,
```
nice idea!

patrickvonplaten commented on src/diffusers/image_processor.py:
```diff
@@ -52,8 +52,16 @@ def __init__(
     ):
         super().__init__()
+        if do_convert_rgb and do_convert_grayscale:
+            warnings.warn(
```
maybe better to throw an error here actually

patrickvonplaten commented:
Very nice first draft! cc @sayakpaul @pcuenca @williamberman here for a review as well

sayakpaul commented on src/diffusers/image_processor.py:
```diff
@@ -51,9 +51,18 @@ def __init__(
     ):
         super().__init__()
+        if do_convert_rgb and do_convert_grayscale:
+            raise ValueError(
+                "`do_convert_rgb` and `do_convert_grayscale` can not both be set to `True`,"
+                " if you intended to convert the image into RGB format, please set `do_convert_grayscale = False`.",
+                " if you intended to convert the image into grayscale format, please set `do_convert_rgb = False`",
+            )
```
Very descriptive! Looks good.

sayakpaul commented on src/diffusers/image_processor.py:
```diff
@@ -119,31 +128,84 @@ def denormalize(images):
+        `vae_scale_factor`.
+
+        Args:
+            image(`PIL.Image.Image`, `np.ndarray` or `torch.Tensor`):
+                the image input, can be a PIL image, numpy array or pytorch tensor. if it is a numpy array, should have
+                shape [batch, height, width] or [batch, height, width, channel] if it is a pytorch tensor, should have
+                shape [batch, channel, height, width]
+            height (`int`, *optional*, defaults to `None`):
+                The height in preprocessed image. If `None`, will use the height of `image` input
+            width (`int`, *optional*`, defaults to `None`):
+                The width in preprocessed. If `None`, will use the width of the `image` input
```
Let's maintain casing:

* "the" -> "The" when it's being placed at the beginning. 
* Fullstops to end the sentences.

sayakpaul commented on src/diffusers/image_processor.py:
```diff
@@ -119,31 +128,84 @@ def denormalize(images):
+        """
+        create a mask
+        """
+        image[image < 0.5] = 0
+        image[image >= 0.5] = 1
+        return image
```
Should 0 and 1 be registered into the config vars?

sayakpaul commented on src/diffusers/image_processor.py:
```diff
@@ -154,6 +216,25 @@ def preprocess(
+
+        # Expand the missing dimension for 3-dimensional pytorch tensor or numpy array that represents grayscale image
+        if self.config.do_convert_grayscale and isinstance(image, (torch.Tensor, np.ndarray)) and image.ndim == 3:
+            if isinstance(image, torch.Tensor):
+                # if image is a pytorch tensor could have 2 possible shapes:
+                #    1. batch x height x width: we should insert the channel dimension at position 1
+                #    2. channnel x height x width: we should insert batch dimension at position 0,
+                #       however, since both channel and batch dimension has same size 1, it is same to insert at position 1
+                #    for simplicity, we insert a dimension of size 1 at position 1 for both cases
+                image = image.unsqueeze(1)
+            else:
+                # if it is a numpy array, it could have 2 possible shapes:
+                #   1. batch x height x width: insert channel dimension on last position
+                #   2. height x width x channel: insert batch dimension on first position
+                if image.shape[-1] == 1:
+                    image = np.expand_dims(image, axis=0)
+                else:
+                    image = np.expand_dims(image, axis=-1)
+
```
For easier operability, does it make sense to first convert the input tensor to a NumPy array and then operate from there?

yiyixuxu replied:
@sayakpaul 

Our logic to convert the PyTorch tensor into a NumPy array assumes a 3D tensor has shape `channel x height x width`, which doesn't apply to grayscale images

for example if we have a tensor with shape `(5, 256,256)`, if we try to convert it into numpy with current logic directly, we will do this:
 `(5, 256,256)` -> put it into a list :`[ (5,256,256)]` -> torch.stack the list when not 4d: `(1, 5, 256, 256)`

sayakpaul replied:
Why cannot we add a dummy channel to represent the grayscale images then? Just trying to understand it better.

yiyixuxu replied:
@sayakpaul 
yeah and that's what this section of code is doing: 
Adding a dummy channel to represent either a missing channel or batch dimension to remove the ambiguity

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:
```diff
@@ -858,9 +866,14 @@ def __call__(
+
+        mask = self.mask_processor.preprocess(mask_image, height=height, width=width)
+
+        masked_image = init_image * (mask < 0.5)
```
Register 0.5 as a config var?

patrickvonplaten replied:
Do you mean of the init of the VAEProcessor or the vae model config ? IMO it's general enough to not have to be registered in a config

sayakpaul reviewed (commented):
Looking quite good. 

Let's try to think of a non-exhaustive list of the scenarios we need to think of to ensure this refactor is robust and write test cases for each of them.

yiyixuxu commented:
@patrickvonplaten 
should we unify the default behavior for when the user does not pass the `height` and `width` arguments?

in SD inpaint pipeline, when `height` and `width` is `None`, it will default to `sample_size * vae scale factor`;
in controlnet inpaint, it will resize to the closest multiple of `8`, which I think makes much more sense

https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py#L866

yiyixuxu commented on src/diffusers/image_processor.py:
```diff
@@ -164,42 +249,47 @@ def preprocess(
         # expected range [0,1], normalize to [-1,1]
         do_normalize = self.config.do_normalize
-        if image.min() < 0:
+        if image.min() < 0 and do_normalize:
```
if user configured the image_processor to have `do_normalize=False`, the expected range should be `[-1,1]` and we shouldn't send a warning (this is the case for controlnet image)

patrickvonplaten commented on src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py:
```diff
@@ -943,8 +925,22 @@ def _encode_vae_image(self, image: torch.Tensor, generator: torch.Generator):
         prompt: Union[str, List[str]] = None,
-        image: Union[torch.Tensor, PIL.Image.Image] = None,
-        mask_image: Union[torch.Tensor, PIL.Image.Image] = None,
+        image: Union[
+            torch.FloatTensor,
+            PIL.Image.Image,
+            np.ndarray,
+            List[torch.FloatTensor],
+            List[PIL.Image.Image],
+            List[np.ndarray],
+        ] = None,
```
```suggestion
        image: Union[
            torch.FloatTensor,
            PIL.Image.Image,
            np.ndarray,
            List[torch.FloatTensor],
            List[PIL.Image.Image],
            List[np.ndarray],
        ] = None,
```
might be worth to define a new type here actually

patrickvonplaten reviewed (commented):
Looks cool! Can we apply the changes also right away to SDXL and SDXL controlnet?

yiyixuxu commented:
>  Can we apply the changes also right away to SDXL and SDXL controlnet?

updated SDXL-inpainting

SDXL controlnet does not have masks input, and we already used a image_processor to process the control condition:) we've added Image_processor to all the other SDXL pipelines wherever there is image input

yiyixuxu commented:
don't now why this test fail - it passes on my machine and this PR did not touch the paint_by_example pipeline at all https://github.com/huggingface/diffusers/actions/runs/5959195493/job/16165075058?pr=4444#step:6:13213

patrickvonplaten reviewed (approved):
Very nice! Let's merge it before we run into ugly merge conflicts :-)
